### ISSUE TITLE
[Bug?] CSS/JS compress kills backend (white screen of death)?

### ISSUE THREAD

**Issue by exside** [modxcms/revolution#12478]
### Summary
Since around 2 weeks I experience a very strange problem with the MODX manager being completely blank (or in some cases only the top menu loaded) after login, the issue is found around the compress CSS/JS system settings, but the definite cause is not clear yet...I had extensive talks with my hoster but besides enabling PHP 5.5s opcache nothing seems to have changed in recent weeks/months, turning off opcache doesn't solve the problem though.

### Step to reproduce
Enable compress_js and compress_css system settings in a 2.3.x installation and try to login...yeah...usually that works, I know =)...I cannot really do something to reproduce, just all my 2.3.x sites have the same issue without me doing anything on these sites (was in holidays for several weeks...)

### Observed behavior
White screen of death upon login into the manger

### Expected behavior
Normal working backend admin interface =)

### Environment
MODX 2.3.x, LiteSpeed webserver (like apache, running since January this year without any problems), mysql 5 something, all browsers

This is seriously a very strange thing and I'm still just guessing what the problem is. To circle it: It's clear that the minify library located at manger/min/ will not deliver anything until the request times out (have timeout at 300s usually and as the manger makes several requests to min this easily adds up to like 30mins...and after that nothing works as no js/css is delivered), the question is why suddenly now? Nothing changed on the code for 2 years (see repo) and nothing changed on the server setup that I know of (except the enabling of opcache, see above)...so I tried to debug as good as I could and found that the problem is not the actual minify library but the two following lines in manger/min/index.php:

```php
$modx->initialize('mgr');
if (!$modx->user->hasSessionContext('mgr')) die();
```

if I comment these out, everything works again as expected...also the ->initialize('mgr') line seems to be fine, just the call to $modx->user->something() seems to fail and indeed when logging just the $modx object, the user property is empty...but really, no idea where this is coming from...looks a bit like the assets are requested before the $modx->user object is set and has a session context, so it seems a check for the user object would help, something like that:

```php
$modx->initialize('mgr');
if (!$modx->user) {
    $modx->getUser();
}
if (!$modx->user->hasSessionContext('mgr')) {
    die();
}
```

Does anything speak against a change like that? If not, I'd make a PR out of this...

**Comment by jpdevries** [comment]
All the more reason to default these settings to off. Not only do they break certain environments, they are anti-optimal
http://forums.modx.com/thread/91496/why-compress-css-and-compress-js-are-silly

**Comment by exside** [comment]
agree with that completely and still I'd like to understand, what is going on =)

**Comment by amdbuilder** [comment]
This is hardly a new problem, in our case we just gave up trying to solve it.  Two identical servers one would have the issue, the other wouldn't have the issue.  I'm all for eliminating these settings, and perhaps getting a refund on the hours of time spent troubleshooting sites because of them.

Yes, I am bias and feel these settings should die a very public painful death.  Granted, JP also makes some good points.

**Comment by amdbuilder** [comment]
See also: #12388, #11142, #10077, #7815

**Comment by jpdevries** [comment]
Especially when considering the possibility of sharing assets across /managers using a common CDN, the myth of server side compression being optimal is very easy to debunk. Regardless of that though, it's almost as non-sensical to ask a server to concat and minify front end files as it is to _wait for it_ do so. 

Your latency becomes relevant to how fast and how taxed your server is. There may be instances where it is faster to server the individual static files then it is to wait for the server to prep them and then send them. Your server is probably busy serving up front end pages to users as it should be! No need to slow it down processing manager files like this.

Also, get this: If you pass in a file path that doesn't exist to the minifer it __completely breaks everything__.

**Comment by OptimusCrime** [comment]
#11682 is an attempt to solve these issues by minifying everything when building.

**Comment by salscode** [comment]
@jpdevries if you want to go the CDN route, cdnjs may be worth looking into more. Shouldn't be impossible to get the MODX manager assets added and CloudFlare hosts them for free. To add/update libs in cdnjs someone just has to submit a pull request on the cdnjs repo in GitHub.

Option 2 (the not so nice option): A paid CDN is used. The cost for this could be quite high as bandwidth is not cheap. There are many CDN providers that could be used but at the end of the day, someone will have to pay the cost.

**Comment by jpdevries** [comment]
@salscode I kind of doubt we'll see CDN routing in MODX 2.x but I do think it should be proposed for 3.x to at least load any common libraries (like Angular, React or whatever else). If that happens, we could submit any MODX specific assets to cdnjs.

I think it would be important to try and keep things working offline so something like the h5bp trick of detect then `document.write` could be used
https://github.com/h5bp/html5-boilerplate/blob/master/dist/index.html#L26

**Comment by adamwintle** [comment]
A nice idea, also remember most CDNs don't work in mainland China and other such territories. There should always be an option to fall back to self-hosted assets.

**Comment by jpdevries** [comment]
> A nice idea, also remember most CDNs don't work in mainland China and other such territories. There should always be an option to fall back to self-hosted assets.

Agreed. This is how h5bp loads jQuery with fallback detection

```html
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="js/vendor/jquery-1.11.3.min.js"><\/script>')</script>
```

I think something similar would be best practice
 - try and load from a CDN
 - detect if it worked
 - local fallback if needed

We'd also have "offline mode" support to consider and this should work for that too

**Comment by adamwintle** [comment]
Also note that in some places, such as China, the CDN version will continuously attempt to load, or start loading but never finish and not a complete hard fail. This means the fallback detection never kicks in because the script believes the CDN version is still downloading. 

So I'd suggest an setting to have complete "offline mode"/keeping assets on local server.

**Comment by amdbuilder** [comment]
Perhaps a better approach would be to add a few system settings to handle this problem.  One to toggle CDN on/off, and then one per resource allowing users to set their prefered CDN.

@adamwintle Do you know if http://www.jsdelivr.com/ is affected?  I recall you noting Google and CDNJS are affected.

**Comment by Mark-H** [comment]
Going all the way back to the original topic off this issue... 
> Does anything speak against a change like that? If not, I'd make a PR out of this...

@exside if adding the getUser() line there resolves the problem for you, I definitely have no issue accepting a pull request that adds it. That said, it shouldn't be necessary because the `$modx->initialize('mgr')` line already calls `$modx->getUser()` via `$modx->_initSession()`...

What strikes me as odd as well is that you say it times out. If it was an actual error (e.g. $modx->user is not a modUser object causing a fatal "undefined is not an object" error) it wouldn't actually time out; it would just return a blank response near instantly..

**Comment by exside** [comment]
yes, that's what keeps me from making this PR, I'm still not sure wtf is going on here =)...

**Comment by pixelchutes** [comment]
See #12538

**Comment by Mark-H** [comment]
Closed per #12611
